galley/pkg/config/processing/snapshotter/snapshotter.go

```diff
@@ -0,0 +1,181 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package snapshotter
+
+import (
+	"time"
+
+	"istio.io/istio/galley/pkg/config/collection"
+	"istio.io/istio/galley/pkg/config/event"
+	"istio.io/istio/galley/pkg/config/processing/snapshotter/strategy"
+	"istio.io/istio/galley/pkg/runtime/monitoring"
+)
+
+// Snapshotter is a processor that handles input events and creates snapshotImpl collections.
+type Snapshotter struct {
+	accumulators map[collection.Name]*accumulator
+	selector     *event.Selector
+	xforms       []event.Transformer
+	settings     []SnapshotOptions
+
+	// lastEventTime records the last time an event was received.
+	lastEventTime time.Time
+
+	// pendingEvents counts the number of events awaiting publishing.
+	pendingEvents int64
+
+	// lastSnapshotTime records the last time a snapshotImpl was published.
+	lastSnapshotTime time.Time
+}
+
+var _ event.Processor = &Snapshotter{}
+
+// HandlerFn handles generated snapshots
+type HandlerFn func(*collection.Set)
+
+type accumulator struct {
+	reqSyncCount int
+	syncCount    int
+	collection   *collection.Instance
+	strategies   []strategy.Instance
+}
+
+// Handle implements event.Handler
+func (a *accumulator) Handle(e event.Event) {
+	switch e.Kind {
+	case event.Added, event.Updated:
+		a.collection.Set(e.Entry)
+		monitoring.RecordStateTypeCount(e.Source.String(), a.collection.Size())
+	case event.Deleted:
+		a.collection.Remove(e.Entry.Metadata.Name)
+		monitoring.RecordStateTypeCount(e.Source.String(), a.collection.Size())
+	case event.FullSync:
+		a.syncCount++
+	default:
+		scope.Errorf("accumulator.Handle: unhandled event type: %v", e.Kind)
+		return
+	}
+
+	if a.syncCount >= a.reqSyncCount {
+		for _, s := range a.strategies {
+			s.OnChange()
+		}
+	}
+}
+
+func (a *accumulator) reset() {
+	a.syncCount = 0
+	a.collection.Clear()
+}
+
+// NewSnapshotter returns a new Snapshotter.
+func NewSnapshotter(xforms []event.Transformer, settings []SnapshotOptions) *Snapshotter {
+	s := &Snapshotter{
+		accumulators:  make(map[collection.Name]*accumulator),
+		selector:      event.NewSelector(),
+		xforms:        xforms,
+		settings:      settings,
+		lastEventTime: time.Now(),
+	}
+
+	for _, xform := range xforms {
+		for _, i := range xform.Inputs() {
+			s.selector.Select(i, xform)
+		}
+
+		for _, o := range xform.Outputs() {
+			a, found := s.accumulators[o]
+			if !found {
+				a = &accumulator{
+					collection: collection.New(o),
+				}
+				s.accumulators[o] = a
+			}
+			a.reqSyncCount++
+			xform.Select(o, a)
+		}
+	}
+
+	for _, o := range settings {
+		for _, c := range o.Collections {
+			a := s.accumulators[c]
+			if a == nil {
+				scope.Errorf("NewSnapshotter: Unrecognized collection in SnapshotOptions: %v (Group: %s)", c, o.Group)
```

Comment:
panic?

Reply:
actually, error is probably a better approach here. The method is publicly exposed from the package. These get parsed from a metadata file. Ideally the sanity checks on the file should be good enough, but an error is probably more appropriate, in case it is not.